src/main/java/com/sapient/model/service/BudgetCategoryService.java

```diff
@@ -0,0 +1,124 @@
+        }
+
+        BudgetCategory budgetCategory = new BudgetCategory();
+
+        Budget budget = budgetService.getBudget(passwordHash, budgetId);
+
+        budgetCategory.setCategory(categoryService.getCategory(passwordHash, categoryId));
+        budgetCategory.setBudget(budget);
+        budgetCategory.setAmount(amount);
+        budgetCategory.setUser(user);
+
+        budgetCategoryDao.save(budgetCategory);
+        return budgetCategory;
+    }
+
+    public BudgetCategory updateBudgetCategory(String passwordHash, Integer id, Float amount) throws NotAuthorizedException {
+        User user;
+        BudgetCategory budgetCategory;
+        try{
+            user = userService.getUserByPasswordHash(passwordHash);
+            budgetCategory = getBudgetCategory(passwordHash, id);
+        }catch (UserNotFoundException | RecordNotFoundException e){
+            throw new NotAuthorizedException("You are not authorized to update this budgetCategory");
+        }
+        budgetCategory.setAmount(amount);
+
+        budgetCategoryDao.save(budgetCategory);
+        return budgetCategory;
+    }
+
+    public void deleteBudgetCategory(String passwordHash, Integer id) throws RecordNotFoundException, NotAuthorizedException {
+        if (!budgetCategoryExists(id)) {
+            throw new RecordNotFoundException("BudgetCategory not found with id: "+ id);
+        }
+        BudgetCategory budgetCategory = getBudgetCategory(passwordHash, id);
+        budgetCategoryDao.delete(budgetCategory);
+    }
+
+    public BudgetCategory getBudgetCategory(String passwordHash, Integer id) throws RecordNotFoundException, NotAuthorizedException {
+        BudgetCategory budgetCategory = budgetCategoryDao.findById(id).orElse(null);
+        if(budgetCategory==null){
+            throw new RecordNotFoundException();
+        }
+        if(!budgetCategory.getUser().getPasswordHash().equals(passwordHash)){
+            throw new NotAuthorizedException("You are not authorized to access this budgetCategory");
+        }
+        return budgetCategory;
+    }
+
+    public List<BudgetCategory> getBudgetCategories(String passwordHash) throws NotAuthorizedException {
+        User user;
+        try {
+            user = userService.getUserByPasswordHash(passwordHash);
+        }catch(Exception e){
+            throw new NotAuthorizedException("Invalid passwordHash");
+        }
+
+        List<BudgetCategory> budgetCategories = new ArrayList<BudgetCategory>();
+
+        for(BudgetCategory budgetCategory: budgetCategoryDao.findAll()){
+            if(budgetCategory.getUser().getId() == user.getId()){
+                budgetCategories.add(budgetCategory);
+            }
+        }
+        return budgetCategories;
+    }
+
+    public Boolean budgetCategoryTaken(Integer categoryId, Integer budgetId, Integer userId){
+        for(BudgetCategory budgetCategory: budgetCategoryDao.findAll()){

```

Comment:
You have the primary key of the users table. Get the user, time complexity of O(1), then get the user's budget categories.